--- 1/src/gui/embedded/qmousetslib_qws.cpp	2014-04-10 22:37:12.000000000 +0400
+++ 2/src/gui/embedded/qmousetslib_qws.cpp	2016-05-18 16:12:06.932368276 +0300
@@ -254,7 +254,7 @@
         int dy = sample.y - lastSample.y;
 
         // Remove small movements in oppsite direction
-        if (dx * lastdx < 0 && qAbs(dx) < jitter_limit) {
+        /*if (dx * lastdx < 0 && qAbs(dx) < jitter_limit) {
             sample.x = lastSample.x;
             dx = 0;
         }
@@ -264,7 +264,7 @@
         }
 
         if (wasPressed == pressed && dx == 0 && dy == 0)
-            return;
+            return;*/
 
 #ifdef TSLIBMOUSEHANDLER_DEBUG
         qDebug() << "last" << QPoint(lastSample.x, lastSample.y)
