--- 1/src/gui/embedded/qscreenlinuxfb_qws.cpp	2016-09-15 13:39:15.681004805 +0300
+++ 2/src/gui/embedded/qscreenlinuxfb_qws.cpp	2016-09-15 13:39:21.307995414 +0300
@@ -1386,6 +1386,18 @@
     return true;
 }
 
+void QLinuxFbScreen::blit(const QImage &image, const QPoint &topLeft,
+	const QRegion &region)
+{
+    ioctl(d_ptr->fd, FBIO_WAITFORVSYNC, NULL);
+    QScreen::blit(image, topLeft, region);
+}
+void QLinuxFbScreen::solidFill(const QColor &color, const QRegion &region)
+{
+    ioctl(d_ptr->fd, FBIO_WAITFORVSYNC, NULL);
+    QScreen::solidFill(color, region);
+}
+
 QT_END_NAMESPACE
 
 #endif // QT_NO_QWS_LINUXFB
--- 1/src/gui/embedded/qscreenlinuxfb_qws.h	2016-09-15 13:39:09.739014721 +0300
+++ 2/src/gui/embedded/qscreenlinuxfb_qws.h	2016-09-15 13:39:21.307995414 +0300
@@ -92,6 +92,8 @@
 
     virtual void disconnect();
     virtual void shutdownDevice();
+    void blit(const QImage &image, const QPoint &topLeft, const QRegion &region);
+    void solidFill(const QColor &color, const QRegion &region);
     virtual void setMode(int,int,int);
     virtual void save();
     virtual void restore();
