diff --git a/mcc_linux.c b/mcc_linux.c
index cabf257..7b037c9 100644
--- a/mcc_linux.c
+++ b/mcc_linux.c
@@ -773,7 +773,7 @@ static int __init mcc_init(void) /* Constructor */
 
 		np = of_find_node_by_name(NULL, "cpu2cpu");
 		irq = of_irq_get(np, count);
-		if ((res = devm_request_irq(dev, irq, cpu_to_cpu_irq_handler, 0, MCC_DRIVER_NAME, mscm_base)) != 0)
+		if ((res = devm_request_irq(dev, irq, cpu_to_cpu_irq_handler, IRQF_NO_SUSPEND, MCC_DRIVER_NAME, mscm_base)) != 0)
 		{
 			printk(KERN_ERR "Failed to register MVF CPU TO CPU interrupt:%d\n",count);
 
