Index: libvo/vo_fbdev.c
===================================================================
diff --git a/libvo/vo_fbdev.c b/libvo/vo_fbdev.c
--- a/libvo/vo_fbdev.c	(revision 37608)
+++ b/libvo/vo_fbdev.c	(working copy)
@@ -1030,7 +1030,7 @@
     if (vidix_name)
         return vidix_query_fourcc(format);
 #endif
-    if (fb_omap && fb_vinfo.nonstd) {
+    if (fb_vinfo.nonstd) {
         if ((fb_vinfo.nonstd == 1 && format == IMGFMT_UYVY) ||
             (fb_vinfo.nonstd == 8 && format == IMGFMT_YUY2))
             return VFCAP_ACCEPT_STRIDE | VFCAP_CSP_SUPPORTED | VFCAP_CSP_SUPPORTED_BY_HW;
