diff --git a/libbb/read_key.c b/libbb/read_key.c
index ace23de..afc8254 100644
--- a/libbb/read_key.c
+++ b/libbb/read_key.c
@@ -119,6 +119,12 @@ int64_t FAST_FUNC read_key(int fd, char *buffer, int timeout)
 				errno = EAGAIN;
 				return -1;
 			}
+
+			/* dk: Handle tty hangups and errors */
+			if (pfd.revents & (POLLERR | POLLHUP)) {
+				errno = ECONNABORTED;
+				return -1;
+			}
 		}
 		/* It is tempting to read more than one byte here,
 		 * but it breaks pasting. Example: at shell prompt,
