diff -uNr busybox-1.23.1.orig/libbb/speed_table.c busybox-1.23.1/libbb/speed_table.c
--- busybox-1.23.1.orig/libbb/speed_table.c	2016-02-01 10:32:30.796931454 +0300
+++ busybox-1.23.1/libbb/speed_table.c	2016-02-01 17:31:21.636290389 +0300
@@ -16,7 +16,7 @@
 #else
 	unsigned short speed;
 #endif
-	unsigned short value;
+	unsigned value;
 };
 
 static const struct speed_map speeds[] = {
@@ -40,24 +40,30 @@
 	{EXTA, 19200},
 #endif
 #ifdef B38400
-	{B38400, 38400/256 + 0x8000U},
+	{B38400, 38400},
 #elif defined(EXTB)
-	{EXTB, 38400/256 + 0x8000U},
+	{EXTB, 38400},
 #endif
 #ifdef B57600
-	{B57600, 57600/256 + 0x8000U},
+	{B57600, 57600},
 #endif
 #ifdef B115200
-	{B115200, 115200/256 + 0x8000U},
+	{B115200, 115200},
 #endif
 #ifdef B230400
-	{B230400, 230400/256 + 0x8000U},
+	{B230400, 230400},
 #endif
 #ifdef B460800
-	{B460800, 460800/256 + 0x8000U},
+	{B460800, 460800},
 #endif
 #ifdef B921600
-	{B921600, 921600/256 + 0x8000U},
+	{B921600, 921600},
+#endif
+#ifdef B1000000
+	{B1000000, 1000000},
+#endif
+#ifdef B1152000
+	{B1152000, 1152000},
 #endif
 };
 
@@ -69,9 +75,6 @@
 
 	do {
 		if (speed == speeds[i].speed) {
-			if (speeds[i].value & 0x8000U) {
-				return ((unsigned long) (speeds[i].value) & 0x7fffU) * 256;
-			}
 			return speeds[i].value;
 		}
 	} while (++i < NUM_SPEEDS);
